zio/ndjsonio/typeparser.go

```diff
@@ -0,0 +1,315 @@
+		return nil, 0, err
+	}
+
+	builder := zcode.NewBuilder()
+	colno := 0
+	nestedColno := 0
+	nestedColumns := info.descriptor.Columns
+	flatColumns := info.flatDesc.Columns
+	for i := range flatColumns {
+		val := jsonVals[i].val
+		if i == info.descriptor.TsCol {
+			ts, err := parseJSONTimestamp(val)
+			if err != nil {
+				return nil, 0, err
+			}
+			if ts < 0 {
+				return nil, 0, fmt.Errorf("negative ts")
+			}
+		}
+
+		recType, isRecord := nestedColumns[colno].Type.(*zng.TypeRecord)
+		if isRecord && nestedColno == 0 {
+			builder.BeginContainer()
+		}
+
+		switch jsonVals[i].typ {
+		case jsonparser.Array:
+			builder.BeginContainer()
+			ztyp := zng.InnerType(flatColumns[i].Type)
+			var iterErr error
+			callback := func(v []byte, typ jsonparser.ValueType, offset int, _ error) {
+				zv, err := parseSimpleType(v, ztyp)
+				if err != nil {
+					iterErr = err
+				} else {
+					builder.AppendPrimitive(zv)
+				}
+			}
+			if _, err := jsonparser.ArrayEach(val, callback); err != nil {
+				return nil, 0, err
+			}
+			if iterErr != nil {
+				return nil, 0, iterErr
+			}
+			builder.EndContainer()
+		case jsonparser.NotExist, jsonparser.Null:
+			switch flatColumns[i].Type.(type) {
+			case *zng.TypeSet, *zng.TypeArray:
+				builder.AppendContainer(nil)
+			default:
+				builder.AppendPrimitive(nil)
+			}
+		default:
+			zv, err := parseSimpleType(val, flatColumns[i].Type)
+			if err != nil {
+				return nil, 0, err
+			}
+			builder.AppendPrimitive(zv)
+		}
+
+		if isRecord {
+			nestedColno += 1
+			if nestedColno == len(recType.Columns) {
+				builder.EndContainer()
+				nestedColno = 0
+				colno += 1
+			}
+		} else {
+			colno += 1
+		}
+
+	}
+	return builder.Bytes(), droppedFields, nil
+}
+
+func (p *typeParser) parseObject(b []byte) (zng.Value, error) {
+
+	var lineNo int
+	incr := func(stat *int) {
+		(*stat)++
+		if p.stats.FirstBadLine == 0 {
+			p.stats.FirstBadLine = lineNo
+		}
+	}
+
+	lineNo++
+	// calling findTypeInfo for each line is costly and we
+	// could optimize for the common case where a
+	// single-type log is posted.
+	ti, err := findTypeInfo(p.zctx, b, p.tr, p.defaultPath)
+	if err != nil {
+		switch err {
+		case ErrDescriptorNotFound:
+			incr(&p.stats.JSONDescriptorNotFound)
+		case ErrMissingPath:
+			incr(&p.stats.JSONMissingPath)
+		default:
+			panic("unhandled error")
+		}
+		return zng.Value{}, err
+	}
+	if ti.flatDesc.TsCol < 0 {
+		incr(&p.stats.BadFormat)
+		return zng.Value{}, ErrBadFormat
+	}
+
+	raw, dropped, err := ti.newRawFromJSON(b)
+	if err != nil {
+		incr(&p.stats.BadFormat)
+		return zng.Value{}, ErrBadFormat
+	}
+	if dropped > 0 {
+		incr(&p.stats.JSONIncompleteDescriptor)
+	}
+
+	return zng.Value{ti.descriptor, raw}, nil
+}
+
+func parseSimpleType(value []byte, typ zng.Type) ([]byte, error) {
+	switch typ {
+	case zng.TypeTime:
+		ts, err := parseJSONTimestamp(value)
+		if err != nil {
+			return nil, err
+		}
+		return zng.EncodeTime(ts), nil
```

Comment:
Nit: You can shorten all of these cases by leaving out the if.
```suggestion
		ts, err := parseJSONTimestamp(value)
		return zng.EncodeTime(ts), nil
```

Reply:
@nwt what am i missing here... the suggested change ignores the err returned by parseJSONTimestamp?

Reply:
Doh! It sure does. Here's what I meant to suggest.
```suggestion
		ts, err := parseJSONTimestamp(value)
		return zng.EncodeTime(ts), err
```

Reply:
Ah ok. I like the reduction in lines, and it would work just fine, but I think it's a bit clearer the current way, as opposed to calling a subsequent function including with an invalid value.